compiler/rustc_codegen_ssa/src/mir/mod.rs

```diff
@@ -260,6 +260,31 @@ pub fn codegen_mir<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     // Apply debuginfo to the newly allocated locals.
     fx.debug_introduce_locals(&mut start_bx);
 
+    if mir.args_iter().any(|arg| {
+        // Find the monomorphized type of our argument
+        let arg_decl = &mir.local_decls[arg];
+        let arg_ty = fx.monomorphize(arg_decl.ty);
+
+        // And check if it is a reference to an uninhabited type
+        let ty = arg_ty.peel_refs();
+        start_bx.layout_of(ty).abi.is_uninhabited()
+    }) {
+        // We're probably not supposed to mutate MIR here.
+        // But since we have an arena, we can just create a clone and replace all references to
+        // the old MIR with the new.
+        let mut body = mir.clone();
+        let blocks = body.basic_blocks.as_mut();
+        blocks.truncate(1);
+        let block = &mut blocks[rustc_middle::mir::START_BLOCK];
+        block.statements.clear();
+        block.terminator.as_mut().unwrap().kind = rustc_middle::mir::TerminatorKind::Unreachable;
+        block.is_cleanup = false;
+
+        // Now we have new MIR, stash it in the TyCtxt arena and swap it in
+        mir = fx.cx.tcx().arena.alloc(body);
+        fx.mir = mir;
```

Comment:
cg_clif has this code which directly emits an unreachable instruction instead for uninhabited arguments: https://github.com/bjorn3/rustc_codegen_cranelift/blob/c07d1e2f88cb3b1a0604ae8f18b478c1aeb7a7fa/src/base.rs#L252-L261 In any case would you mind adding the `peel_refs()` call here too if this experiment is a success?